### PR TITLE
client-gen: stop embedding of GroupVersion client intfs

### PR DESCRIPTION
**Pull request by sttts**
It is undefined (or at least uncontrollable) which methods of the clientset apigroup
interfaces are actually inherited. Moreover, there might be nameconflicts between the
accessors and inherited methods. This PR removes the embedding to make it unambiguous.

```release-note
Enforce explicit references to API group client interfaces in clientsets to avoid ambiguity.
```

### CODE REVIEW

**Comment by gmarek** [comment]
+100 for this PR:)

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -485,10 +485,11 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) (err error) {
 			eventClientConfig := *clientConfig
 			eventClientConfig.QPS = float32(s.EventRecordQPS)
 			eventClientConfig.Burst = int(s.EventBurst)
-			eventClient, err = clientgoclientset.NewForConfig(&eventClientConfig)
+			client, err := clientgoclientset.NewForConfig(&eventClientConfig)
 			if err != nil {
 				glog.Warningf("Failed to create API Server client: %v", err)
 			}
+			eventClient = client.CoreV1()
```
Comment (ash2k) [inline cmd/kubelet/app/server.go]: This will panic with npe?

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -413,7 +413,13 @@ func createClients(config componentconfig.ClientConnectionConfiguration, masterO
 		return nil, nil, err
 	}
 
+<<<<<<< HEAD
```
Comment (sttts) [inline cmd/kube-proxy/app/server.go]: rebase conflicts...

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (sttts) [inline cmd/kubelet/app/server.go]: fixed

**Comment by caesarxuchao** [comment]
Could you make a release note since it's a non-compatibility change?

File: staging/src/k8s.io/kube-gen/cmd/client-gen/generators/generator_for_clientset.go [hunk 3]
```diff
@@ -69,8 +69,15 @@ func (g *genClientset) GenerateType(c *generator.Context, t *types.Type, w io.Wr
 	// perhaps we can adapt the go2ild framework to this kind of usage.
 	sw := generator.NewSnippetWriter(w, c, "$", "$")
 
-	allGroups := clientgentypes.ToGroupVersionPackages(g.groups)
+	type groupArg struct {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/kube-gen/cmd/client-gen/generators/generator_for_clientset.go]: Could you just add `LowerCaseGroupVersion` to `GroupVersionPackage` and update `ToGroupVersionPackages`

File: pkg/kubectl/versioned_client.go [hunk 4]
```diff
@@ -18,28 +18,13 @@ package kubectl
...
-
-func versionedClientsetForDaemonSet(internalClient internalclientset.Interface) externalclientset.Interface {
-	if internalClient == nil {
-		return &externalclientset.Clientset{}
-	}
-	return &externalclientset.Clientset{
-		AppsV1beta1Client:       apps.New(internalClient.Apps().RESTClient()),
-		ExtensionsV1beta1Client: extensions.New(internalClient.Extensions().RESTClient()),
-	}
+	return externalclientset.New(internalClient.Core().RESTClient())
```
Comment (caesarxuchao) [inline pkg/kubectl/versioned_client.go]: Does this work? I think `internalClient.Core().RESTClient()` and `internalClient.Apps().RESTClient()` are targeting at different URL.

Comment (sttts) [inline pkg/kubectl/versioned_client.go]: Good point. Fixed. I really disliked half-nil clientsets. That's bad practice. I have switched all the code using `versionedClientset` to explicit, versioned api group clientset interfaces. Much cleaner IMO now.

Comment (caesarxuchao) [inline pkg/kubectl/versioned_client.go]: Yes, it's much better.

File: staging/src/k8s.io/kube-gen/cmd/client-gen/generators/generator_for_clientset.go (hunk 3 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/kube-gen/cmd/client-gen/generators/generator_for_clientset.go]: @sttts how about this comment?

Comment (sttts) [inline staging/src/k8s.io/kube-gen/cmd/client-gen/generators/generator_for_clientset.go]: fixed

**Comment by sttts** [comment]
@gmarek @caesarxuchao finally fixed all implicit casts, all green. ptal

File: pkg/printers/internalversion/describe.go [hunk 5]
```diff
@@ -3633,14 +3632,20 @@ func (list SortableVolumeMounts) Less(i, j int) bool {
 	return list[i].MountPath < list[j].MountPath
 }
 
-func versionedClientsetForDeployment(internalClient clientset.Interface) versionedclientset.Interface {
+// TODO: get rid of this and plumb the caller correctly
+func versionedClientset(internalClient clientset.Interface) versionedclientset.Interface {
```
Comment (caesarxuchao) [inline pkg/printers/internalversion/describe.go]: Remove the function if no one is using it. If it's still being used, update it to versionedCoreClientV1.

Comment (sttts) [inline pkg/printers/internalversion/describe.go]: Wasn't used anymore. Fixed.

**Comment by sttts** [comment]
Added changelog note. @caesarxuchao ptal

**Comment by sttts** [comment]
Unit test flake is fixed in https://github.com/kubernetes/kubernetes/pull/50092

**Pull request merged**